bigtop-deploy/puppet/modules/spark/templates/spark-defaults.conf

```diff
@@ -13,7 +13,15 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+<% if @master_url -%>
 spark.master <%= @master_url %>
+<% else -%>
+<% if scope['deploy::roles'].include? 'spark-on-yarn' -%>
+spark.master yarn
+<% else -%>
+spark.master spark://<%= @master_host %>:<%= @master_port %>
```

Comment:
I think this can be thoroughly discussed because backward compatibility is indeed important.

I did it this way just because the logic deploying spark (spark-master, spark-worker) implies it's a standalone deployment. However it happens to be working spark on yarn instead.

How about this way:
spark remains deploying yarn, and then I'll add another component call spark-standalone:
The code change will be looked like this:

```
spark => {
  worker => ["spark-on-yarn"],
  client => ["spark-client"],
  library => ["spark-yarn-slave"],
}
spark-standalone => {
  master => ["spark-master"],
  worker => ["spark-worker"],
}
```